### ISSUE TITLE
Question

### ISSUE THREAD

**Issue by gelsas** [ashevchuk/midea-ac#1]
Hello,

I have just found your repository and wanted to ask if your script is working via LAN (no cloud etc. needed?)?

(I also have found this repository: https://github.com/mac-zhou/midea-msmart, I think he tires the same as you. Maybe you can take a look if there is something you can add to your script aswell)

**Comment by ashevchuk** [comment]
That's right, the script works over a local network, without a cloud, etc.
Thanks for the link to a similar project.
I will definitely look at the capabilities of this project and add the missing features to my project.
Thanks again.

**Comment by gelsas** [comment]
Great to hear! Will try out the script in the next few days! 

Do you know the home automation software https://www.openhab.org/ ?
If yes is there any chance that it could be integrated with openhab ?

**Comment by ashevchuk** [comment]
Yes, I know about the "openhab" project. But I have never used it.
The script for controlling the air conditioner, I made it as close as possible to the Unix philosophy.
It is a simple command line interface to the air conditioner and makes it very easy to use it in any other software system.
As far as I understand, to create a binding for an "openhab", I need to do it using the JAVA language?
Unfortunately, I have never used JAVA, but I can try to make a wrapper for an "openhab".

**Comment by gelsas** [comment]
Hi,

Thank you for the quick response! 

Yes Openhab bindings are in Java language.

A wrapper for an "openhab" would be great!

**Comment by gelsas** [comment]
I do have one more question would it be possible to automatically check for midea air conditioners in the LAN ? 
Basically an "auto discovery" feature?

**Comment by ashevchuk** [comment]
The only reliable way of "auto discovery", I'm afraid, is by scanning the range of subnet addresses. Maybe I'm wrong. I can add such a function.

**Comment by gelsas** [comment]
That would be great!

**Comment by gelsas** [comment]
I have just now had time to test your script for the first time. 

I tested this script "ac_nodeps.pl" since I was not sure how to install the missing dependencies on my raspberry pi.

It works like a charm, it does exactly what it should do! Thank you for the great work 🥇 👍 

Is it normal that the command is executed delayed on the AC ?
When I send the command it takes about 20-40 seconds before for example the tempature actually gets changed on the AC.

**Comment by ashevchuk** [comment]
The problem with latency is that "ac_nodeps.pl" uses the "nc" utility, in which you can only specify the timeout, but not the number of bytes received. This timeout value is specified at the beginning of the "ac_nodeps.pl" script itself. It's pretty big just in case. You can change it to your value.
The "ac.pl" script doesn't have this problem, but you need to install several Perl modules.
"ac_nodeps.pl" is itself an experimental script and will be improved in the future.
"ac.pl" is a stable, production quality script.

**Comment by gelsas** [comment]
Great thank you for the quick and detailed response. I already thought something like that might be the case. 

Would it be possible that you could let me know what commands I need to use to install the missing dependencies on my raspbian (raspberry pi4) ?
I would really love to be able to test the ac.pl script.

EDIT: just found out what the name of the needed dependency was: libcryptx-perl was missing on my raspberry pi

**Comment by ashevchuk** [comment]
It looks like installing two packages should be enough (if Perl version 5.28):
```
apt-get install libcryptx-perl libdigest-crc-perl
```
Another installation option:
```
apt-get install cpanminus
cpanm CryptX Digest::CRC
```

**Comment by ashevchuk** [comment]
Added **--discover** switch(ac.pl script) to search for compatible devices on the specified network.
The network can take values in the form of:
single IP address: ```...--discover --ip 192.168.1.1```
network with mask: ```...--discover --ip 192.168.1.0/255.255.255.0```
network with mask length: ```...--discover --ip 192.168.1.0/24```
range of IP addresses: ```...--discover --ip 192.168.1.1-192.168.1.8```

**Comment by gelsas** [comment]
> Added **--discover** switch(ac.pl script) to search for compatible devices on the specified network.
> The network can take values in the form of:
> single IP address: `...--discover --ip 192.168.1.1`
> network with mask: `...--discover --ip 192.168.1.0/255.255.255.0`
> network with mask length: `...--discover --ip 192.168.1.0/24`
> range of IP addresses: `...--discover --ip 192.168.1.1-192.168.1.8`

This works fantastic! Thank you very much for this.

May I ask one more question (hope me asking so much annoying stuff it still okay)?
This guy here: [Midea-MSmart](https://github.com/mac-zhou/midea-msmart) has implemented an auto discovery aswell: "midea-discover", as soon as you run the command it found the device within 1-2 seconds. 
Does he use a different method to detect midea devices?

**Comment by ashevchuk** [comment]
Ask as many questions as necessary. With pleasure, I will try to help.
I looked at how the discovery was done in the "midea-msmart" project and there really was a different approach,
namely sending a broadcast packet and waiting for a response from the devices that detected it.
Unfortunately, my device does not respond to such packets and I was unable to detect it using this method.
It is not clear what can be done about this.
The approach in "ac.pl" script is to check every host that was specified, which is likely to be more reliable.
Also, you can tweak this process a little by changing two values ​​in the script itself.
To do this, you can change two parameters at the beginning of the "ac.pl" script itself: "TIMEOUT" and "SCAN_NPROC_MAX".
The "TIMEOUT" value depends on the quality / reliability / congestion / etc of the network.
The "SCAN_NPROC_MAX" value depends on how many parallel processes you want to use to scan the network.
Of course, the default values ​​are specified as universally as possible and with a margin.
For example, if you specify "TIMEOUT => 1" and "SCAN_NPROC_MAX => 16", then scanning 254 addresses in my network takes 15 seconds.

**Comment by gelsas** [comment]
Thank you for the quick resposne. I have changed the values to the ones you mentioned and now the discovery is also very fast at 15-16 seconds.

On the openhab forum there are hundreds of people who have midea ACs and would love a binding that is why I have thought that I maybe be able to somehow realize this and somehow get it into a binding.

Since I am not a coder I have thought that I maybe  can find someone who could create such binding or a rewritten version of this in Java language on a freelancing platform but I was sadly not able to find someone who was able do it.

Do you maybe know some other platforms I could try to find someone who could rewrite a perl script into Java?

**Comment by ashevchuk** [comment]
Unfortunately, I don’t know of such platforms, as I have never used.
I thought about trying to make the binding myself, as you asked, but I can't say when in time it will be done due to the availability of free time(I will need to understand both the "openhab" and the Java language).

**Comment by gelsas** [comment]
Great to hear you want to try to create such a binding. You are amazing. Thank you for all the work you are putting into this.

In my own research on how to create a binding I have found this documentation
https://www.openhab.org/docs/developer/ and asked in the openhab forum for guidance.
They told me while the documentation there is almost completely up to date, only one information is not completely correct anymore. Java 8 (has as far as I understood reach its "End of Life" in december) which is why the recommended version to use now is Java 11.

Furthermore this advice was given to other people asking about creating bindings aswell: 

> There is a script (in the documentation) that will create a skeleton binding for you to use as a starting point.
> 
> Then take a look at a few existing bindings that do something similar to what yours will need to do. For example, if you are accessing a REST API with JSON payloads over HTTP. look for another binding that accesses a REST API using JSON payloads over HTTP. This will save you a lot of time, since everything you need to know is probably not in the docs.
> 
> There is an XML file that describes the things and channels that your binding will provide. You may want to think that through before you jump in to the actual binding code.
>
>Furthermore the openhab community is always happy to help if you hit any problems while creating the binding and will assist the best way they can.

Hope this information helps somehow. That was what I was able to find out.

Thank you again for all the time and effort you are putting into this. I really appreciate it.
Is there any way to support you somehow via something like https://ko-fi.com/ ?

**Comment by ashevchuk** [comment]
Thanks for the tips. I think it would be logical, for a start, to look at the existing similar bindings.
With regard to material support, there is no need to worry. Thanks a lot for that.
I did it anyway for my own needs. And if it is somehow useful to someone else, I will only be glad.

**Comment by gelsas** [comment]
I just had time to test each function of the script in detail now and there are two commands which do not work: --turbo and --swing. Every other command works exactly as expected/as it should.

Is there anyway that we could find out why those two are not working for me ?

**Comment by ashevchuk** [comment]
Made minor adjustments. Try to see if this resolves the --swing mode problem.
With the --turbo mode, I still need to do some research.

EDIT: Also, possibly fixed --turbo mode. Try this mode on your device as well.

**Comment by ashevchuk** [comment]
I also found a solution to the problem with a quick search for devices on the network.
For fast UDP broadcast searches, use the following parameters and their values: ```...--discover --ip 255.255.255.255```

**Comment by gelsas** [comment]
Thank you for all the changes! 
Unfortunately I am not able to test them whatever command I try I always get the following error:
`unknown "swing" value: "48" at /midealan/ac.pl line 587.`

I was able to test the discovery function and it now takes 2-4 seconds and the ACs are found.

**Comment by ashevchuk** [comment]
This means that your device has encoded values that do not exist in my device. This is fixable. A little later, I'll make a correction.

**Comment by gelsas** [comment]
Please let me know if I can help in any way to fix it, I will try my best to be as much of assistance as I can be.
If needed I could even probably capture the traffic between app and ac if that is needed to find out correct values or something like that (but I would need guidance on how to do it),

(Please let me know if I ask to much questions or if I am bothering you to much then I will stop with it immediately)

**Comment by ashevchuk** [comment]
It's okay, don't worry about it.
It's just that I've been too busy for the last week.
I just need some time to finish the started changes. With new changes, such problems should not arise in principle.
I think I can finish the changes on Monday, since I'm not at home right now.

**Comment by ashevchuk** [comment]
Sorry, a little delayed, but you can test the latest version of the script.

**Comment by gelsas** [comment]
Thank you for your response.

No need to apologize and no need to rush things. Take all the time you need.
I really appreciate your help!


```
openhab@openhab:~$ /etc/scripts/ac.pl --ip 192.168.1.47 --set --swing both
swing:both
openhab@openhab:~$ /etc/scripts/ac.pl --ip 192.168.1.47 --set --turbo on
turbo:off
openhab@openhab:~$
```


I just tried your latest changes and now the swing feature is working flawlessly. 

The only thing not working for me now, is setting the turbo.
Above you can see the response when sending turbo on. 

Any idea if this is fixable? 

Can I help somehow ?

**Comment by ashevchuk** [comment]
This may be due to a delay in the application of the device operating mode parameters.
Try to set the parameter first:
```/etc/scripts/ac.pl --ip 192.168.1.47 --set --turbo on```
Then, after a few seconds, just read the parameter:
```/etc/scripts/ac.pl --ip 192.168.1.47 --get --turbo```
If the parameter remains unchanged while reading, then we will look further, what could be the problem.

**Comment by gelsas** [comment]
I have just tried as suggested and the parameter remains unchanged

**Comment by ashevchuk** [comment]
Updated the script. You can try to see if these changes fix the turbo mode issue.
I noticed that turning off the turbo mode on my device occurs with some delay.

**Comment by gelsas** [comment]
I just tested the latest version of the script but unfortunately it still does not work. 

What I also just tested was what the --get command shows when I have activated the turbo via the remote. 

It still shows the the turbo status as off. 

Does that information help?
Any suggestions what we could do to get turbo working ? 

Do I need to capture the traffic the app sends or something like that so we can find out how to get it working ?

**Comment by gelsas** [comment]
I have just found this: https://github.com/reneklootwijk/node-mideahvac and I asked there how he is setting the turbo mode compared to your script and that guy said that the difference is in setting the bit. 

He says that in this script it is setting the wrong bit  in the command or something like that. 

Could that be ? Maybe we can take a look at his code to see the difference

**Comment by ashevchuk** [comment]
Thanks for the link. If anything, then I have a manufacturer's application that I can study. Corrected the script again. You can try that out of it. Also, I will later add verbose debug output with protocol decoding that will make it much easier to fix such problems.

**Comment by gelsas** [comment]
Just tried the latest version but sadly the turbo mode is still not working.

I think we need to add "verbose debug output with protocol decoding" as you suggested.
That way we can find out what is going on.

**Comment by reneklootwijk** [comment]
I am not sure your device uses the same protocol version as mine, but I can control turbo mode by (un)setting bit 1 (value 0x02) of byte 10 counted from the byte specifying the command (0x40). It is the same byte where you (un)set bit 2 (value 0x04) to set the temperature unit to celcius or fahrenheit.

**Comment by gelsas** [comment]
Hi @ashevchuk, I hope I am not disturbing/annoying you. I just wanted to ask if you maybe already had time to add the verbose debug option with protocol decoding.

Best, 
gelsas

**Comment by ashevchuk** [comment]
Yes, I was able to add an option for debugging. Also, I decided to repeat, one by one, the code for setting and parsing the required bit flags, exactly as in the manufacturer's application. In a couple of days, I will add another script to the repository - this script will be used to debug problems.

**Comment by gelsas** [comment]
Great, thank you for the update on the progress and thank you for helping me out so much!

**Comment by sergge1** [comment]
Hello!
Does this git support devices that use protocol 3 for communication?
I have an midea like ac.
I have bought EU-SK103X wifi dongle and ac works well through the midea cloud (with app on my phone).

However I can't control it locally. 
Asi realized, due to the protocol 3.

**Comment by gelsas** [comment]
Hi @ashevchuk , did you already had the time to finish the other script ?
Best,
gelsas

**Comment by gelsas** [comment]
Hi @ashevchuk hope everything is okay with you.

I have found that the turbo mode is set successfully when using msmart (https://github.com/mac-zhou/midea-msmart/blob/9e178befd6ac70b52a4b7e82124db9c9da7aac5b/msmart/command.py)
What do I need to change in your code to match here how the turbo mode is set in msmart?

best,
gelsas

**Comment by ashevchuk** [comment]
It's been a while (the war makes adjustments), but I added the --debug mode.
There were some minor protocol fixes. Update the script.
After the commands
./ac.pl --ip xxxx --set --turbo on
try running the command
./ac.pl --ip xxxx --get --turbo
after a while, since changing some operating modes doesn't happen instantly and takes some time.
